### PR TITLE
Add fonts to support more different types of characters for multiple languages

### LINKED ISSUE #29495: [Heartbeat] Missing some i18n fonts in heartbeat docker image
We've had reports that some characters (Korean) don't show up when using synthetics via the docker image (heartbeat and agent-complete). The most likely cause is missing fonts. We should add as many fonts as we can (within reason) and validate that we can visit sites in various languages we don't now test in (Korean, Arabic, etc.)

### PR DESCRIPTION
**Pull request by lucasfcosta**
> ⚠️ Please let me know if we should backport this to 7.17.

## What does this PR do?

This PR fixes #29495 by adding more fonts to support multiple different characters from multiple languages.

After upgrading the base image, even though all characters from a few languages worked fine, not _all_ characters from _all_ languages were being displayed.

Therefore, this PR installs the following packages within the `elastic-agent` and `heartbeat` images so that those can work fine:

```
fonts-arphic-ukai
fonts-arphic-uming
fonts-ipafont-mincho
fonts-ipafont-gothic
fonts-unfonts-core
```

## Why is it important?

This is important because customer's websites use a wide variety of language and we must have an inclusive synthetics product capable of displaying a wide variety of characters in the screenshots it takes.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

I'd appreciate it if reviewers could check the following items:

- [ ] Do these dependencies cover all characters we're planning on supporting?
- [ ] Are there more font-packs we need? If so, which?
- [ ] Does this need a backport for 7.17?

## For Heartbeat

1. Checkout this branch
2. Navigate into the `heartbeat` directory within `x-pack`
3. Build its image using `env PLATFORMS="+all linux/amd64" mage package`.
4. Create a file with monitors which access websites in different languages:
    ```yaml
    output.elasticsearch:
      hosts: ["http://elasticsearch:9200"]
      username: "elastic"
      password: "changeme"

    heartbeat.monitors:
    - type: browser
      id: korean-characters
      name: korean_characters_test
      schedule: '@every 1m'
      source:
        inline:
          script: |-
            step("first", async () => {
              await page.goto('https://www.naver.com/');
            });
    - type: browser
      id: chinese-characters
      name: chinese_characters_test
      schedule: '@every 1m'
      source:
        inline:
          script: |-
            step("first", async () => {
              await page.goto('https://baike.baidu.com/');
            });
    - type: browser
      id: russian-characters
      name: russian_characters_test
      schedule: '@every 1m'
      source:
        inline:
          script: |-
            step("first", async () => {
              await page.goto('https://yandex.ru/');
            });
    - type: browser
      id: arabic-characters
      name: arabic_characters_test
      schedule: '@every 1m'
      source:
        inline:
          script: |-
            step("first", async () => {
              await page.goto('https://www.bbc.com/arabic');
            });
    ```
4. Run `elastic-package stack up` and then run the Heartbeat container mounting the configuration file above. That will cause the container to use the locally built image.
    ```
   docker run \
      -it \
      -u 0 \
      --entrypoint=/bin/bash \
      --rm \
      --name=heartbeat \
      --user=heartbeat \
      --volume="/tmp/hb1.yml:/usr/share/heartbeat/heartbeat.yml:ro" \
      --network="elastic-package-stack_default" \
      docker.elastic.co/beats/heartbeat:8.1.0
    ```
5. Check the screenshot which Heartbeat is going to send for each inline monitor.
    
### For Elastic Agent

1. Checkout this branch
2. Navigate into the `elastic-agent` directory within `x-pack`
3. Build its image using `env PLATFORMS="+all linux/amd64" mage dev:package`.
4. Start an Elastic Stack using `elastic-package stack up` using a compatible stack version
5. In Kibana, create browser monitors which access the following websites:
    * https://www.naver.com
    * https://baike.baidu.com
    * https://yandex.ru/
    * https://www.bbc.com/arabic
    An inline script like the one below for each of these websites should be enough:
    ```
    step("first", async () => {
      await page.goto('https://www.naver.com/');
    });
    ```
6. Run a container with the `elastic-agent-complete` image and change its entry-point to `/bin/bash` so you can go into it.
    ```
    docker run \
      -it \
      --entrypoint=/bin/bash \
      --rm \
      --name=elastic-agent \
      --user=elastic-agent \
      --volume="/Users/lucasfcosta/Repositories/synthetics:/usr/share/synthetics" \
      --network="elastic-package-stack_default" \
      docker.elastic.co/beats/elastic-agent-complete:8.1.0
    ```
7. Enroll your agent with `elastic-agent enroll --insecure --url=http://fleet-server:8220 --enrollment-token=POLICY_TOKEN_HERE`. You can get the enrollment token and policy tokens in Kibana (under Fleet).
8. Start your agent with `elastic-agent &` and wait for it to send screenshots for the monitors you've set up in Kibana.

## Related issues

- Fixes #29495

## Screenshots

Screenshot 2022-01-14 at 14 31 41
Screenshot 2022-01-14 at 14 32 05
Screenshot 2022-01-14 at 14 31 51
Screenshot 2022-01-14 at 14 32 15

### CODE REVIEW

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (vigneshshanmugam) [inline dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:73]: I am curios about the list we arrived at here, We don't even have the basic one `fonts-liberation` for the image.

Do we know if that works? 

Here is the list from PW  for comparison - https://github.com/microsoft/playwright/blob/5ba7903ba098586a13745e0d7ac894f1d55d47aa/packages/playwright-core/src/utils/nativeDeps.ts#L434-L470

One we use in RUM agent for puppeteer and playwright - https://github.com/elastic/apm-agent-rum-js/tree/master/.ci/docker

**Review by andrewvc — APPROVED** [review]
I'm LGTM but let's address @vigneshshanmugam 's comments before merging

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (lucasfcosta) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:73]: Here I followed an approach of testing websites and taking note of which didn't display all characters correctly. Then, I went after the packages only for those, and ran an E2E test which generated the images above. I followed this process to avoid adding too many unnecessary packages.

Wrt to fonts-liberation, it seemed to be a standard set of fonts which could just replace the default set of fonts to which it's compatible, but we already have default fonts for western characters in the base image and thus I thought we wouldn't need that. Looking into the other packages, they also seem to fall into one of two categories:
* Covering standard character sets (like `fonts-liberation` or `ttf-unifont`)
* Tools for managing fonts themselves (like `libfontconfig`)

The other fonts seemed to be the most complete _and_ popular fonts for their character standards (`fonts-arphic-ukai` [for chinese characters] , `fonts-ipafont-mincho` [for japanese characters, `fonts-ipafont-gothic` being for sans-serif japanese characters], and `fonts-unfonts-core` for Korean characters). Those were the ones which came up when I looked for packages for each individual language.

We already had support for Arabic/Russian characters in the base image as per the initial test, so I haven't added those as the websites I visited already displayed correctly. It could be the case that there are characters outside the unicode space 

Looking at Playwright's deps these are the changes I see we could make:
* [`fonts-wqy-zenhei`](https://packages.debian.org/sid/fonts-wqy-zenhei) this seems like a more complete font set to replace the one in the package we have. However, I was just thinking about the licensing of these given that [this package seems to be GPL licensed](http://wenq.org/wqy2/index.cgi?ZenHei(en)) and therefore I don't know whether we could include it in the image we distribute. The [Arphic Public License](https://www.freedesktop.org/wiki/Arphic_Public_License/) for the current fonts seem to only apply for modification of the fonts themselves and is incompatible with GPL and thus seems fine, but actually it would be good to review.
    ➡️  @andrewvc @vigneshshanmugam do we have a particular specialist in software licensing to review these?
* [`fonts-tlwg`](https://linux.thai.net/projects/fonts-tlwg) seems to support Thai characters, so that's one I think we could add. [However, it also includes GPL licensed fonts, so we must double check that](https://github.com/tlwg/fonts-tlwg/blob/master/COPYING).
* [`fonts-noto-color-emoji`] for supporting Google's emoji set (which should really follow the same set as the standard, shouldn't it?)

I moved this PR to `draft` just so that we would make sure to review the licensing and whether we're happy with the current coverage set. I do thing we should add `fonts-tlwg` though, so I'll run a test for that before opening this PR for review again too.

**Comment by lucasfcosta** [comment]
After talking about this issue with @vigneshshanmugam, I've addressed a couple more of his points:

* I double-checked that **the new image has good support for default fallback fonts**. I did so by accessing different websites with all requests enabled and then accessing the very same websites while disabling font requests in playwright.
  Here is an example of a script I used.
  ```js
  step("cnn without fonts", async () => {
      await page.route('**/*', (route) => {
            return route.request().resourceType() === 'font'
                ? route.abort()
                : route.continue()
      });
      await page.goto('https://cnn.com');
  });
  
  step("cnn with fonts", async () => {
      await page.route('**/*', (route) => route.continue());
      await page.goto('https://cnn.com');
  });
  ```
  I also accessed https://furbo.org/stuff/systemfonts-new.html, which has default font values in CSS and loads no font files. In that website, all of the pieces of text rendered fine in the system's fallback font (although they don't have stylistic differences as we're using a fallback).
* Furthermore, I also confirmed that we _cannot_ run headful tests in the new `ubuntu` images because there's no X display in them (as shown in the error log below).
    ```
    heartbeat@0699f71daf07:~$ synthetics /usr/share/synth-example/ --no-headless
      Failed to run the test browserType.launch: Protocol error (Browser.getVersion): Browser closed.
      ==================== Browser output: ====================
      <launching> /usr/share/heartbeat/.cache/ms-playwright/chromium-907428/chrome-linux/chrome --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-breakpad --disable-client-side-phishing-detection --disable-component-extensions-with-background-pages --disable-default-apps --disable-dev-shm-usage --disable-extensions --disable-features=ImprovedCookieControls,LazyFrameLoading,GlobalMediaControls,DestroyProfileOnBrowserClose --allow-pre-commit-input --disable-hang-monitor --disable-ipc-flooding-protection --disable-popup-blocking --disable-prompt-on-repost --disable-renderer-backgrounding --disable-sync --force-color-profile=srgb --metrics-recording-only --no-first-run --enable-automation --password-store=basic --use-mock-keychain --no-service-autorun --user-data-dir=/tmp/playwright_chromiumdev_profile-vWKBgW --remote-debugging-pipe --no-sandbox --no-startup-window
      <launched> pid=244
      [pid=244][err] [244:244:0119/160156.284241:ERROR:browser_main_loop.cc(1400)] Unable to open X display.
      =========================== logs ===========================
      <launching> /usr/share/heartbeat/.cache/ms-playwright/chromium-907428/chrome-linux/chrome --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-breakpad --disable-client-side-phishing-detection --disable-component-extensions-with-background-pages --disable-default-apps --disable-dev-shm-usage --disable-extensions --disable-features=ImprovedCookieControls,LazyFrameLoading,GlobalMediaControls,DestroyProfileOnBrowserClose --allow-pre-commit-input --disable-hang-monitor --disable-ipc-flooding-protection --disable-popup-blocking --disable-prompt-on-repost --disable-renderer-backgrounding --disable-sync --force-color-profile=srgb --metrics-recording-only --no-first-run --enable-automation --password-store=basic --use-mock-keychain --no-service-autorun --user-data-dir=/tmp/playwright_chromiumdev_profile-vWKBgW --remote-debugging-pipe --no-sandbox --no-startup-window
      <launched> pid=244
      [pid=244][err] [244:244:0119/160156.284241:ERROR:browser_main_loop.cc(1400)] Unable to open X display.
      ============================================================
          at Function.setupDriver (/usr/share/heartbeat/.node/node/lib/node_modules/@elastic/synthetics/src/core/gatherer.ts:48:43)
          at Function.createContext (/usr/share/heartbeat/.node/node/lib/node_modules/@elastic/synthetics/src/core/runner.ts:157:35)
          at Runner.runJourney (/usr/share/heartbeat/.node/node/lib/node_modules/@elastic/synthetics/src/core/runner.ts:410:34)
          at Runner.run (/usr/share/heartbeat/.node/node/lib/node_modules/@elastic/synthetics/src/core/runner.ts:488:22)
          at Object.run (/usr/share/heartbeat/.node/node/lib/node_modules/@elastic/synthetics/src/index.ts:45:12)
          at /usr/share/heartbeat/.node/node/lib/node_modules/@elastic/synthetics/src/cli.ts:201:19 {
        name: 'Error'
      }
    ```
    It's also worth highlighting that we were also not able to run headful tests on the old Heartbeat image too. I tried running Headful tests within the `7.16` image and got the same error.
    @vigneshshanmugam do you think that should be a supported case within the images? I feel like it's up to the user to install X and create the necessary display if they'd like to run headful tests, but I'm of course open to discussing it.
    Also, if we do end-up supporting headful runs, that means we will _not_ disable GPU acceleration, which may cause failures inside these containers, so we'd need to handle that better with `libmesa` or one of the alternatives we discussed earlier.

**Comment by vigneshshanmugam** [comment]
@lucasfcosta Thanks Lucas for testing it and providing the details. 

>  In that website, all of the pieces of text rendered fine in the system's fallback font 

Great. 

> do you think that should be a supported case within the images? I feel like it's up to the user to install X and create the necessary display if they'd like to run headful tests, but I'm of course open to discussing it.

Interesting, I didn't know the old image was not able to run headful tests. I am fine with any of the below options
- Leave it up to the user to do the full setup.
- We can add xvfb to the dependencies and let the user do the rest. 

> Also, if we do end-up supporting headful runs, that means we will not disable GPU acceleration, which may cause failures inside these containers, so we'd need to handle that better with libmesa or one of the alternatives we discussed earlier.

Gotcha, we should make a call here.  Lets see what others have to say @andrewvc @paulb-elastic @drewpost

**Comment by drewpost** [comment]
> @lucasfcosta Thanks Lucas for testing it and providing the details.
> 
> > In that website, all of the pieces of text rendered fine in the system's fallback font
> 
> Great.
> 
> > do you think that should be a supported case within the images? I feel like it's up to the user to install X and create the necessary display if they'd like to run headful tests, but I'm of course open to discussing it.
> 
> Interesting, I didn't know the old image was not able to run headful tests. I am fine with any of the below options
> 
> * Leave it up to the user to do the full setup.
> * We can add xvfb to the dependencies and let the user do the rest.
> 
> > Also, if we do end-up supporting headful runs, that means we will not disable GPU acceleration, which may cause failures inside these containers, so we'd need to handle that better with libmesa or one of the alternatives we discussed earlier.
> 
> Gotcha, we should make a call here. Lets see what others have to say @andrewvc @paulb-elastic @drewpost

My gut feel is that it's OK to disable GPU acceleration. We're not a device simulation testing product. What's key is to keep the results consistent and as long as we pick one option or the other and stick with it, then that requirement is met. There may be real world impact on the sizing we need to allow for given the more inefficient CPU only approach and I'm open to that as being something we discuss from an operation efficiency standpoint but from a product perspective, I'm OK with not enabling GPU acceleration. cc @lucasfcosta @vigneshshanmugam

**Comment by lucasfcosta** [comment]
Hey @drewpost thanks a lot for your input! 😊  With regards to your last point about GPU acceleration I've recently run some tests within containers and it seems like containers do not have access to the underlying host's GPU so there's no actual difference in performance.

I've added a few more details in [this link](https://github.com/elastic/synthetics/pull/441#issuecomment-1014379777).

**Comment by andrewvc** [comment]
I've pushed a change to use google's SIL licensed noto fonts. All the tests listed previously still work, with all languages rendering.

**Review by andrewvc — APPROVED** [review]
LGTM

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (vigneshshanmugam) [inline dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:79]: Why do we explicitly install the fonts on elastic-agent image and not on the HB image?

**Review by vigneshshanmugam — APPROVED** [review]
LGTM with pending comment https://github.com/elastic/beats/pull/29861#discussion_r793888435

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (andrewvc) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:79]: ah, that's a bug, it's actually installed higher up, this is redundant

**Comment by andrewvc** [comment]
@Mergifyio backport 7.17

**Pull request merged**
